src/tests/ftest/cart/util/daos_sys_logscan.py

```diff
@@ -0,0 +1,561 @@
+    ldr_status_re = r"rebuild_leader_status_check\(\).*" + rbid_re + r" \[(\w+)\].*duration=(\d+)"
+    re_rebuild_ldr_start = re.compile(ldr_start_re)
+    re_rebuild_ldr_status = re.compile(ldr_status_re)
+
+    # Legacy rebuild PS leader logging (before uniform rebuild string)
+    old_ldr_start_re = r"rebuild_leader_start.*([0-9a-fA-F]{8}).*version=(\d+)/(\d+).*op=(\w+)"
+    old_ldr_status_re = (r"rebuild_leader_status_check\(\) (\w+) \[(\w+)\] \(pool ([0-9a-fA-F]{8}) "
+                         r"leader (\d+) term (\d+).*ver=(\d+),gen (\d+).*duration=(\d+) secs")
+    re_old_ldr_start = re.compile(old_ldr_start_re)
+    re_old_ldr_status = re.compile(old_ldr_status_re)
+
+    def __init__(self):
+        # dictionaries indexed by pool UUID
+        self._pools = {}
+        self._highest_pmapver = {}
+        self._cur_ldr_rank = {}
+        self._cur_ldr_pid = {}
+        self._cur_term = {}
+
+        # filename to rank map
+        self._file_to_rank = {}
+
+        self._warnings = []
+        self._check_rb_new_fmt = True
+        self._check_rb_legacy_fmt = True
+        self._debug = False
+
+    def _warn(self, wmsg, fname, line=None):
+        full_msg = f"WARN file={fname}"
+        if line:
+            full_msg += f", line={line.lineno}"
+        full_msg += f": {wmsg}"
+
+        self._warnings.append(full_msg)
+        print(full_msg)
+
+    def get_warnings(self):
+        """Return all warnings stored when scanning engine log files"""
+        return self._warnings
+
+    def _set_rank(self, log_iter):
+        print(f"INFO: searching for rank in file {log_iter.fname}")
+        found = False
+        for line in log_iter.new_iter():
+            # when a rank assignment log line found (engine start)
+            match = self.re_rank_assign.match(line.get_msg())
+            if match:
+                self._file_to_rank[log_iter.fname] = int(match.group(1))
+                found = True
+                break
+
+            # Future enhancement: what about log rotation (not an engine start scenario)?
+        return found
+
+    # return log-message, hostname, and date/timestamp components of the line
+    def _get_line_components(self, line):
+        return line.get_msg(), line.hostname, line.time_stamp
+
+    # is this rank, pid the leader of the pool with uuid puuid?
+    def _is_leader(self, puuid, rank, pid):
+        if puuid not in self._pools:
+            return False
+        if self._cur_ldr_rank[puuid] == rank and self._cur_ldr_pid[puuid] == pid:
+            return True
+        return False
+
+    def _get_ps_leader_components(self, match):
+        # puuid, term
+        # see re_step_up and re_step_down
+        return match.group(1), int(match.group(2))
+
+    def _match_ps_step_up(self, fname, line, pid, rank):
+        msg, host, datetime = self._get_line_components(line)
+        match = self.re_step_up.match(msg)
+        if match:
+            puuid, term = self._get_ps_leader_components(match)
```

Comment:
I think better to reverse this `if` and avoid large nesting
```suggestion
        match = self.re_step_up.match(msg)
        if not match:
            return False

        puuid, term = self._get_ps_leader_components(match)
```

Reply:
done in  latest commit